File: QolmsJotoWebView/QolmsJotoWebView/Workers/HealthAgeWorker.cs
using MGF.QOLMS.QolmsApiCoreV1;
using MGF.QOLMS.QolmsApiEntityV1;
using MGF.QOLMS.QolmsJotoWebView.Repositories;
using System;
using System.Linq;

namespace MGF.QOLMS.QolmsJotoWebView
{
    /// <summary>
    /// 「健康年齢」画面に関する機能を提供します。
    /// このクラスは継承できません。
    /// </summary>
    /// <remarks></remarks>
    internal sealed class HealthAgeWorker
    {

        IHealthAgeRepository _healthAgeRepo;

        #region Constant

        private const int DEFAULT_DATA_COUNT = 3;

        #endregion

        #region Constructor

        /// <summary>
        /// デフォルトコンストラクタは使用できません。
        /// </summary>
        /// <remarks></remarks>
        public HealthAgeWorker(IHealthAgeRepository healthAgeRepo)
        {
            _healthAgeRepo = healthAgeRepo;
        }

        #endregion

        #region Private Method


        #endregion

        #region Public Method

        /// <summary>
        /// 「健康年齢」画面ビューモデルを作成します。
        /// </summary>
        /// <param name="mainModel">メインモデル。</param>
        /// <param name="fromPageNoType">遷移元の画面番号の種別。</param>
        /// <returns>
        /// 「健康年齢」画面ビューモデル。
        /// </returns>
        /// <remarks></remarks>
        public HealthAgeViewModel CreateViewModel(QolmsJotoModel mainModel, QjPageNoTypeEnum fromPageNoType)
        {
            HealthAgeViewModel result = mainModel.GetInputModelCache<HealthAgeViewModel>(); // キャッシュから取得

            if (result == null || result.MembershipType != mainModel.AuthorAccount.MembershipType || result.FromPageNoType != fromPageNoType)
            {
                // キャッシュが無ければ API を実行
                QhYappliHealthAgeReadApiResults apiResult = _healthAgeRepo.ExecuteHealthAgeReadApi(mainModel);

                result = new HealthAgeViewModel(mainModel, fromPageNoType);

                // TODO:
                result.AvailableHealthAgeReportN = apiResult.AvailableHealthAgeReportN.ConvertAll(
                    i => new AvailableHealthAgeReportItem()
                    {
                        HealthAgeReportType = i.HealthAgeReportType.TryToValueType(QjHealthAgeReportTypeEnum.None),
                        LatestDate = i.LatestDate.TryToValueType(DateTime.MinValue)
                    }
                );

                // 会員の種別
                result.MembershipType = apiResult.MembershipType.TryToValueType(QjMemberShipTypeEnum.Free);

                // アカウント情報の会員の種別を更新
                mainModel.AuthorAccount.MembershipType = result.MembershipType;
                // result.MembershipType = QjMemberShipTypeEnum.Premium; // TODO: debug

                // 病院連携中かのフラグ
                result.HasHospitalConnection = apiResult.LinkegeSystemNoN.Any();

                // 健康年齢
                result.HealthAge = apiResult.HealthAge.TryToValueType(decimal.MinValue);

                // 実年齢との差
                result.HealthAgeDistance = apiResult.HealthAgeDistance.TryToValueType(decimal.MinValue);

                // 測定日
                result.LatestDate = apiResult.LatestDate.TryToValueType(DateTime.MinValue);

                // 予測医療費
                result.MedicalExpenses = apiResult.MedicalExpenses.TryToValueType(int.MinValue);

                // 健康年齢の推移のリスト
                result.HealthAgeN = apiResult.HealthAgeN.ConvertAll(i => i.ToHealthAgeValueItem());

                // 健康年齢改善アドバイス情報のリスト
                result.HealthAgeAdviceN = apiResult.HealthAgeAdviceN.ConvertAll(i => i.ToHealthAgeAdviceItem());

                // キャッシュへ追加
                mainModel.SetInputModelCache(result);
            }

            return result;
        }

        public QjHealthAgeReportPartialViewModelBase CreateReportPartialViewModel(
            QolmsJotoModel mainModel,
            HealthAgeViewModel pageViewModel,
            QjHealthAgeReportTypeEnum healthAgeReportType,
            ref string refPartialViewName)
        {
            refPartialViewName = string.Empty;

            QjHealthAgeReportPartialViewModelBase result = null;

            // 年齢分布
            if (healthAgeReportType == QjHealthAgeReportTypeEnum.Distribution)
            {
                refPartialViewName = "_HealthAgeDistributionReportPartialView";

                bool hasData = false;

                if (pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, ref hasData) && hasData)
                {
                    if (!pageViewModel.DistributionPartialViewModel.ReportItem.HealthAgeValueN.Any())
                    {
                        QhYappliHealthAgeReportReadApiResults reportResult = _healthAgeRepo.ExecuteHealthAgeReportRead(
                            mainModel,
                            QjHealthAgeReportTypeEnum.Distribution,
                            HealthAgeWorker.DEFAULT_DATA_COUNT
                        );

                        result = new HealthAgeDistributionReportPartialViewModel(
                            pageViewModel,
                            reportResult.HealthAgeReport.ToHealthAgeReportItem()
                        ); // TODO:

                        pageViewModel.DistributionPartialViewModel = (HealthAgeDistributionReportPartialViewModel)result;

                        mainModel.SetInputModelCache(pageViewModel);

                        return result;
                    }
                    else
                    {
                        return pageViewModel.DistributionPartialViewModel;
                    }
                }
                else
                {
                    return new HealthAgeDistributionReportPartialViewModel(pageViewModel, null);
                }
            }

            // 肥満
            if (healthAgeReportType == QjHealthAgeReportTypeEnum.Fat)
            {
                refPartialViewName = "_HealthAgeFatReportPartialView";

                bool hasData = false;

                if (pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, ref hasData) && hasData)
                {
                    if (!pageViewModel.FatPartialViewModel.ReportItem.HealthAgeValueN.Any())
                    {
                        QhYappliHealthAgeReportReadApiResults reportResult = _healthAgeRepo.ExecuteHealthAgeReportRead(
                            mainModel,
                            QjHealthAgeReportTypeEnum.Fat,
                            HealthAgeWorker.DEFAULT_DATA_COUNT
                        );

                        result = new HealthAgeFatReportPartialViewModel(
                            pageViewModel,
                            reportResult.HealthAgeReport.ToHealthAgeReportItem()
                        ); // TODO:

                        pageViewModel.FatPartialViewModel = (HealthAgeFatReportPartialViewModel)result;

                        mainModel.SetInputModelCache(pageViewModel);

                        return result;
                    }
                    else
                    {
                        return pageViewModel.FatPartialViewModel;
                    }
                }
                else
                {
                    return new HealthAgeFatReportPartialViewModel(pageViewModel, null);
                }
            }

            // 血糖
            if (healthAgeReportType == QjHealthAgeReportTypeEnum.Glucose)
            {
                refPartialViewName = "_HealthAgeGlucoseReportPartialView";

                bool hasData = false;

                if (pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, ref hasData) && hasData)
                {
                    if (!pageViewModel.GlucosePartialViewModel.ReportItem.HealthAgeValueN.Any())
                    {
                        QhYappliHealthAgeReportReadApiResults reportResult = _healthAgeRepo.ExecuteHealthAgeReportRead(
                            mainModel,
                            QjHealthAgeReportTypeEnum.Glucose,
                            HealthAgeWorker.DEFAULT_DATA_COUNT
                        );

                        result = new HealthAgeGlucoseReportPartialViewModel(
                            pageViewModel,
                            reportResult.HealthAgeReport.ToHealthAgeReportItem()
                        ); // TODO:

                        pageViewModel.GlucosePartialViewModel = (HealthAgeGlucoseReportPartialViewModel)result;

                        mainModel.SetInputModelCache(pageViewModel);

                        return result;
                    }
                    else
                    {
                        return pageViewModel.GlucosePartialViewModel;
                    }
                }
                else
                {
                    return new HealthAgeGlucoseReportPartialViewModel(pageViewModel, null);
                }
            }

            // 血圧
            if (healthAgeReportType == QjHealthAgeReportTypeEnum.Pressure)
            {
                refPartialViewName = "_HealthAgePressureReportPartialView";

                bool hasData = false;

                if (pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, ref hasData) && hasData)
                {
                    if (!pageViewModel.PressurePartialViewModel.ReportItem.HealthAgeValueN.Any())
                    {
                        QhYappliHealthAgeReportReadApiResults reportResult = _healthAgeRepo.ExecuteHealthAgeReportRead(
                            mainModel,
                            QjHealthAgeReportTypeEnum.Pressure,
                            HealthAgeWorker.DEFAULT_DATA_COUNT
                        );

                        result = new HealthAgePressureReportPartialViewModel(
                            pageViewModel,
                            reportResult.HealthAgeReport.ToHealthAgeReportItem()
                        ); // TODO:

                        pageViewModel.PressurePartialViewModel = (HealthAgePressureReportPartialViewModel)result;

                        mainModel.SetInputModelCache(pageViewModel);

                        return result;
                    }
                    else
                    {
                        return pageViewModel.PressurePartialViewModel;
                    }
                }
                else
                {
                    return new HealthAgePressureReportPartialViewModel(pageViewModel, null);
                }
            }

            // 脂質
            if (healthAgeReportType == QjHealthAgeReportTypeEnum.Lipid)
            {
                refPartialViewName = "_HealthAgeLipidReportPartialView";

                bool hasData = false;

                if (pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, ref hasData) && hasData)
                {
                    if (!pageViewModel.LipidPartialViewModel.ReportItem.HealthAgeValueN.Any())
                    {
                        QhYappliHealthAgeReportReadApiResults reportResult = _healthAgeRepo.ExecuteHealthAgeReportRead(
                            mainModel,
                            QjHealthAgeReportTypeEnum.Lipid,
                            HealthAgeWorker.DEFAULT_DATA_COUNT
                        );

                        result = new HealthAgeLipidReportPartialViewModel(
                            pageViewModel,
                            reportResult.HealthAgeReport.ToHealthAgeReportItem()
                        ); // TODO:

                        pageViewModel.LipidPartialViewModel = (HealthAgeLipidReportPartialViewModel)result;

                        mainModel.SetInputModelCache(pageViewModel);

                        return result;
                    }
                    else
                    {
                        return pageViewModel.LipidPartialViewModel;
                    }
                }
                else
                {
                    return new HealthAgeLipidReportPartialViewModel(pageViewModel, null);
                }
            }

            // 肝臓
            if (healthAgeReportType == QjHealthAgeReportTypeEnum.Liver)
            {
                refPartialViewName = "_HealthAgeLiverReportPartialView";

                bool hasData = false;

                if (pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, ref hasData) && hasData)
                {
                    if (!pageViewModel.LiverPartialViewModel.ReportItem.HealthAgeValueN.Any())
                    {
                        QhYappliHealthAgeReportReadApiResults reportResult = _healthAgeRepo.ExecuteHealthAgeReportRead(
                            mainModel,
                            QjHealthAgeReportTypeEnum.Liver,
                            HealthAgeWorker.DEFAULT_DATA_COUNT
                        );

                        result = new HealthAgeLiverReportPartialViewModel(
                            pageViewModel,
                            reportResult.HealthAgeReport.ToHealthAgeReportItem()
                        ); // TODO:

                        pageViewModel.LiverPartialViewModel = (HealthAgeLiverReportPartialViewModel)result;

                        mainModel.SetInputModelCache(pageViewModel);

                        return result;
                    }
                    else
                    {
                        return pageViewModel.LiverPartialViewModel;
                    }
                }
                else
                {
                    return new HealthAgeLiverReportPartialViewModel(pageViewModel, null);
                }
            }

            // 尿糖・尿蛋白
            if (healthAgeReportType == QjHealthAgeReportTypeEnum.Urine)
            {
                refPartialViewName = "_HealthAgeUrineReportPartialView";

                bool hasData = false;

                if (pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, ref hasData) && hasData)
                {
                    if (!pageViewModel.UrinePartialViewModel.ReportItem.HealthAgeValueN.Any())
                    {
                        QhYappliHealthAgeReportReadApiResults reportResult = _healthAgeRepo.ExecuteHealthAgeReportRead(
                            mainModel,
                            QjHealthAgeReportTypeEnum.Urine,
                            HealthAgeWorker.DEFAULT_DATA_COUNT
                        );

                        result = new HealthAgeUrineReportPartialViewModel(
                            pageViewModel,
                            reportResult.HealthAgeReport.ToHealthAgeReportItem()
                        ); // TODO:

                        pageViewModel.UrinePartialViewModel = (HealthAgeUrineReportPartialViewModel)result;

                        mainModel.SetInputModelCache(pageViewModel);

                        return result;
                    }
                    else
                    {
                        return pageViewModel.UrinePartialViewModel;
                    }
                }
                else
                {
                    return new HealthAgeUrineReportPartialViewModel(pageViewModel, null);
                }
            }

            return result;
        }

        #endregion
    }

}

File: QolmsYappli/QolmsYappli/Workers/HealthAgeWorker.vb
Imports MGF.QOLMS.QolmsApiCoreV1
Imports MGF.QOLMS.QolmsApiEntityV1

''' <summary>
''' 「健康年齢」画面に関する機能を提供します。
''' この クラス は継承できません。
''' </summary>
''' <remarks></remarks>
Friend NotInheritable Class HealthAgeWorker

#Region "Constant"

    Private Const DEFAULT_DATA_COUNT As Integer = 3

#End Region

#Region "Constructor"

    ''' <summary>
    ''' デフォルト コンストラクタは使用できません。
    ''' </summary>
    ''' <remarks></remarks>
    Private Sub New()
    End Sub

#End Region

#Region "Private Method"

    Private Shared Function ExecuteHealthAgeReadApi(mainModel As QolmsYappliModel) As QhYappliHealthAgeReadApiResults

        Dim apiArgs As New QhYappliHealthAgeReadApiArgs(
            QhApiTypeEnum.YappliHealthAgeRead,
            QsApiSystemTypeEnum.Qolms,
            mainModel.ApiExecutor,
            mainModel.ApiExecutorName
        ) With {
            .ActorKey = mainModel.AuthorAccount.AccountKey.ToApiGuidString(),
            .PageNo = Convert.ToByte(QyPageNoTypeEnum.HealthAge).ToString(),
            .IsInitialize = Boolean.TrueString,
            .Birthday = mainModel.AuthorAccount.Birthday.ToApiDateString()
        }
        Dim apiResults As QhYappliHealthAgeReadApiResults = QsApiManager.ExecuteQolmsApi(Of QhYappliHealthAgeReadApiResults)(
            apiArgs,
            mainModel.SessionId,
            mainModel.ApiAuthorizeKey
        )

        With apiResults
            If .IsSuccess.TryToValueType(False) Then
                Return apiResults
            Else
                Throw New InvalidOperationException(String.Format("{0} API の実行に失敗しました。", QsApiManager.GetQolmsApiName(apiArgs)))
            End If
        End With

    End Function

    Private Shared Function ExecuteHealthAgeReportRead(mainModel As QolmsYappliModel, healthAgeReportType As QyHealthAgeReportTypeEnum, dataCount As Integer) As QhYappliHealthAgeReportReadApiResults

        Dim apiArgs As New QhYappliHealthAgeReportReadApiArgs(
            QhApiTypeEnum.YappliHealthAgeReportRead,
            QsApiSystemTypeEnum.Qolms,
            mainModel.ApiExecutor,
            mainModel.ApiExecutorName
        ) With {
            .ActorKey = mainModel.AuthorAccount.AccountKey.ToApiGuidString(),
            .HealthAgeReportType = healthAgeReportType.ToString(),
            .DataCount = dataCount.ToString()
        }
        Dim apiResults As QhYappliHealthAgeReportReadApiResults = QsApiManager.ExecuteQolmsApi(Of QhYappliHealthAgeReportReadApiResults)(
            apiArgs,
            mainModel.SessionId,
            mainModel.ApiAuthorizeKey
        )

        With apiResults
            If .IsSuccess.TryToValueType(False) Then
                Return apiResults
            Else
                Throw New InvalidOperationException(String.Format("{0} API の実行に失敗しました。", QsApiManager.GetQolmsApiName(apiArgs)))
            End If
        End With

    End Function

#End Region

#Region "Public Method"

    ''' <summary>
    ''' 「健康年齢」画面 ビュー モデル を作成します。
    ''' </summary>
    ''' <param name="mainModel">メイン モデル。</param>
    ''' <param name="fromPageNoType">遷移元の画面番号の種別。</param>
    ''' <returns>
    ''' 「健康年齢」画面 ビュー モデル。
    ''' </returns>
    ''' <remarks></remarks>
    Public Shared Function CreateViewModel(mainModel As QolmsYappliModel, fromPageNoType As QyPageNoTypeEnum) As HealthAgeViewModel

        Dim result As HealthAgeViewModel = mainModel.GetInputModelCache(Of HealthAgeViewModel)() ' キャッシュから取得

        If result Is Nothing OrElse result.MembershipType <> mainModel.AuthorAccount.MembershipType OrElse result.FromPageNoType <> fromPageNoType Then
            ' キャッシュが無ければ API を実行
            With HealthAgeWorker.ExecuteHealthAgeReadApi(mainModel)
                result = New HealthAgeViewModel(mainModel, fromPageNoType)

                ' TODO:
                result.AvailableHealthAgeReportN = .AvailableHealthAgeReportN.ConvertAll(
                    Function(i)
                        Return New AvailableHealthAgeReportItem() With {
                            .HealthAgeReportType = i.HealthAgeReportType.TryToValueType(QyHealthAgeReportTypeEnum.None),
                            .LatestDate = i.LatestDate.TryToValueType(Date.MinValue)
                        }
                    End Function
                )

                ' 会員の種別
                result.MembershipType = .MembershipType.TryToValueType(QyMemberShipTypeEnum.Free)

                ' アカウント情報の会員の種別を更新
                mainModel.AuthorAccount.MembershipType = result.MembershipType
                'result.MembershipType = QyMemberShipTypeEnum.Premium ' TODO: debug

                ' 病院連携中かの フラグ
                result.HasHospitalConnection = .LinkegeSystemNoN.Any()

                ' 健康年齢
                result.HealthAge = .HealthAge.TryToValueType(Decimal.MinValue)

                ' 実年齢との差
                result.HealthAgeDistance = .HealthAgeDistance.TryToValueType(Decimal.MinValue)

                ' 測定日
                result.LatestDate = .LatestDate.TryToValueType(Date.MinValue)

                ' 予測医療費
                result.MedicalExpenses = .MedicalExpenses.TryToValueType(Integer.MinValue)

                ' 健康年齢の推移のリスト
                result.HealthAgeN = .HealthAgeN.ConvertAll(Function(i) i.ToHealthAgeValueItem())

                ' 健康年齢改善アドバイス情報のリスト
                result.HealthAgeAdviceN = .HealthAgeAdviceN.ConvertAll(Function(i) i.ToHealthAgeAdviceItem())

                ' キャッシュへ追加
                mainModel.SetInputModelCache(result)
            End With
        End If

        Return result

    End Function

    Public Shared Function CreateReportPartialViewModel(mainModel As QolmsYappliModel, pageViewModel As HealthAgeViewModel, healthAgeReportType As QyHealthAgeReportTypeEnum, ByRef refPartialViewName As String) As QyHealthAgeReportPartialViewModelBase

        refPartialViewName = String.Empty

        Dim result As QyHealthAgeReportPartialViewModelBase = Nothing

        ' 年齢分布
        If healthAgeReportType = QyHealthAgeReportTypeEnum.Distribution Then
            refPartialViewName = "_HealthAgeDistributionReportPartialView"

            Dim hasData As Boolean = False

            If pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, hasData) AndAlso hasData Then
                If Not pageViewModel.DistributionPartialViewModel.ReportItem.HealthAgeValueN.Any() Then
                    With HealthAgeWorker.ExecuteHealthAgeReportRead(mainModel, QyHealthAgeReportTypeEnum.Distribution, HealthAgeWorker.DEFAULT_DATA_COUNT)
                        result = New HealthAgeDistributionReportPartialViewModel(
                            pageViewModel,
                            .HealthAgeReport.ToHealthAgeReportItem()
                        ) ' TODO:
                    End With

                    pageViewModel.DistributionPartialViewModel = DirectCast(result, HealthAgeDistributionReportPartialViewModel)

                    mainModel.SetInputModelCache(pageViewModel)

                    Return result
                Else
                    Return pageViewModel.DistributionPartialViewModel
                End If
            Else
                Return New HealthAgeDistributionReportPartialViewModel(pageViewModel, Nothing)
            End If
        End If

        ' 肥満
        If healthAgeReportType = QyHealthAgeReportTypeEnum.Fat Then
            refPartialViewName = "_HealthAgeFatReportPartialView"

            Dim hasData As Boolean = False

            If pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, hasData) AndAlso hasData Then
                If Not pageViewModel.FatPartialViewModel.ReportItem.HealthAgeValueN.Any() Then
                    With HealthAgeWorker.ExecuteHealthAgeReportRead(mainModel, QyHealthAgeReportTypeEnum.Fat, HealthAgeWorker.DEFAULT_DATA_COUNT)
                        result = New HealthAgeFatReportPartialViewModel(
                            pageViewModel,
                            .HealthAgeReport.ToHealthAgeReportItem()
                        ) ' TODO:
                    End With

                    pageViewModel.FatPartialViewModel = DirectCast(result, HealthAgeFatReportPartialViewModel)

                    mainModel.SetInputModelCache(pageViewModel)

                    Return result
                Else
                    Return pageViewModel.FatPartialViewModel
                End If
            Else
                Return New HealthAgeFatReportPartialViewModel(pageViewModel, Nothing)
            End If
        End If

        ' 血糖
        If healthAgeReportType = QyHealthAgeReportTypeEnum.Glucose Then
            refPartialViewName = "_HealthAgeGlucoseReportPartialView"

            Dim hasData As Boolean = False

            If pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, hasData) AndAlso hasData Then
                If Not pageViewModel.GlucosePartialViewModel.ReportItem.HealthAgeValueN.Any() Then
                    With HealthAgeWorker.ExecuteHealthAgeReportRead(mainModel, QyHealthAgeReportTypeEnum.Glucose, HealthAgeWorker.DEFAULT_DATA_COUNT)
                        result = New HealthAgeGlucoseReportPartialViewModel(
                            pageViewModel,
                            .HealthAgeReport.ToHealthAgeReportItem()
                        ) ' TODO:
                    End With

                    pageViewModel.GlucosePartialViewModel = DirectCast(result, HealthAgeGlucoseReportPartialViewModel)

                    mainModel.SetInputModelCache(pageViewModel)

                    Return result
                Else
                    Return pageViewModel.GlucosePartialViewModel
                End If
            Else
                Return New HealthAgeGlucoseReportPartialViewModel(pageViewModel, Nothing)
            End If
        End If

        ' 血圧
        If healthAgeReportType = QyHealthAgeReportTypeEnum.Pressure Then
            refPartialViewName = "_HealthAgePressureReportPartialView"

            Dim hasData As Boolean = False

            If pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, hasData) AndAlso hasData Then
                If Not pageViewModel.PressurePartialViewModel.ReportItem.HealthAgeValueN.Any() Then
                    With HealthAgeWorker.ExecuteHealthAgeReportRead(mainModel, QyHealthAgeReportTypeEnum.Pressure, HealthAgeWorker.DEFAULT_DATA_COUNT)
                        result = New HealthAgePressureReportPartialViewModel(
                            pageViewModel,
                            .HealthAgeReport.ToHealthAgeReportItem()
                        ) ' TODO:
                    End With

                    pageViewModel.PressurePartialViewModel = DirectCast(result, HealthAgePressureReportPartialViewModel)

                    mainModel.SetInputModelCache(pageViewModel)

                    Return result
                Else
                    Return pageViewModel.PressurePartialViewModel
                End If
            Else
                Return New HealthAgePressureReportPartialViewModel(pageViewModel, Nothing)
            End If
        End If

        ' 脂質
        If healthAgeReportType = QyHealthAgeReportTypeEnum.Lipid Then
            refPartialViewName = "_HealthAgeLipidReportPartialView"

            Dim hasData As Boolean = False

            If pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, hasData) AndAlso hasData Then
                If Not pageViewModel.LipidPartialViewModel.ReportItem.HealthAgeValueN.Any() Then
                    With HealthAgeWorker.ExecuteHealthAgeReportRead(mainModel, QyHealthAgeReportTypeEnum.Lipid, HealthAgeWorker.DEFAULT_DATA_COUNT)
                        result = New HealthAgeLipidReportPartialViewModel(
                            pageViewModel,
                            .HealthAgeReport.ToHealthAgeReportItem()
                        ) ' TODO:
                    End With

                    pageViewModel.LipidPartialViewModel = DirectCast(result, HealthAgeLipidReportPartialViewModel)

                    mainModel.SetInputModelCache(pageViewModel)

                    Return result
                Else
                    Return pageViewModel.LipidPartialViewModel
                End If
            Else
                Return New HealthAgeLipidReportPartialViewModel(pageViewModel, Nothing)
            End If
        End If

        ' 肝臓
        If healthAgeReportType = QyHealthAgeReportTypeEnum.Liver Then
            refPartialViewName = "_HealthAgeLiverReportPartialView"

            Dim hasData As Boolean = False

            If pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, hasData) AndAlso hasData Then
                If Not pageViewModel.LiverPartialViewModel.ReportItem.HealthAgeValueN.Any() Then
                    With HealthAgeWorker.ExecuteHealthAgeReportRead(mainModel, QyHealthAgeReportTypeEnum.Liver, HealthAgeWorker.DEFAULT_DATA_COUNT)
                        result = New HealthAgeLiverReportPartialViewModel(
                            pageViewModel,
                            .HealthAgeReport.ToHealthAgeReportItem()
                        ) ' TODO:
                    End With

                    pageViewModel.LiverPartialViewModel = DirectCast(result, HealthAgeLiverReportPartialViewModel)

                    mainModel.SetInputModelCache(pageViewModel)

                    Return result
                Else
                    Return pageViewModel.LiverPartialViewModel
                End If
            Else
                Return New HealthAgeLiverReportPartialViewModel(pageViewModel, Nothing)
            End If
        End If

        ' 尿糖・尿蛋白
        If healthAgeReportType = QyHealthAgeReportTypeEnum.Urine Then
            refPartialViewName = "_HealthAgeUrineReportPartialView"

            Dim hasData As Boolean = False

            If pageViewModel.IsAvailableHealthAgeReportType(healthAgeReportType, hasData) AndAlso hasData Then
                If Not pageViewModel.UrinePartialViewModel.ReportItem.HealthAgeValueN.Any() Then
                    With HealthAgeWorker.ExecuteHealthAgeReportRead(mainModel, QyHealthAgeReportTypeEnum.Urine, HealthAgeWorker.DEFAULT_DATA_COUNT)
                        result = New HealthAgeUrineReportPartialViewModel(
                            pageViewModel,
                            .HealthAgeReport.ToHealthAgeReportItem()
                        ) ' TODO:
                    End With

                    pageViewModel.UrinePartialViewModel = DirectCast(result, HealthAgeUrineReportPartialViewModel)

                    mainModel.SetInputModelCache(pageViewModel)

                    Return result
                Else
                    Return pageViewModel.UrinePartialViewModel
                End If
            Else
                Return New HealthAgeUrineReportPartialViewModel(pageViewModel, Nothing)
            End If
        End If

        Return result

    End Function

#End Region

End Class
